include/aspect/particle/world.h

```diff
@@ -21,729 +21,310 @@
-          AssertThrow (integrator != NULL, ExcMessage ("Particle world integrator must be set before calling init()."));
-
-          // Construct MPI data type for this particle
-          T::add_mpi_types(data_info);
-
-          // And data associated with the integration scheme
-          integrator->add_mpi_types(data_info);
-
-          // Set up the block lengths, indices and internal types
-          num_entries = data_info.size();
-          block_lens = new int[num_entries];
-          indices = new MPI_Aint[num_entries];
-          old_types = new MPI_Datatype[num_entries];
-          for (i=0; i<num_entries; ++i)
-            {
-              block_lens[i] = data_info[i].n_elements;
-              indices[i] = (i == 0 ? 0 : indices[i-1]+sizeof(double)*data_info[i-1].n_elements);
-              old_types[i] = MPI_DOUBLE;
-            }
-
-          // Create and commit the MPI type
-          res = MPI_Type_struct(num_entries, block_lens, indices, old_types, &particle_type);
-          if (res != MPI_SUCCESS) exit(-1);
-
-          res = MPI_Type_commit(&particle_type);
-          if (res != MPI_SUCCESS) exit(-1);
-
-          // Delete temporary arrays
-          delete [] old_types;
-          delete [] indices;
-          delete [] block_lens;
-
-          // Initialize send/recv structures appropriately
-          num_send = new int[world_size];
-          num_recv = new int[world_size];
-          send_offset = new int[world_size];
-          recv_offset = new int[world_size];
-          send_reqs = new MPI_Request[world_size];
-          recv_reqs = new MPI_Request[world_size];
-        };
+        void
+        load_tracers(const typename parallel::distributed::Triangulation<dim>::cell_iterator &cell,
+                     const typename parallel::distributed::Triangulation<dim>::CellStatus status,
+                     const void *data);
 
         /**
-         * Calculate the cells containing each particle for all particles.
+         * Serialize the contents of this class.
          */
-        void find_all_cells()
-        {
-          typename std::multimap<LevelInd, T>::iterator   it;
-          std::multimap<LevelInd, T>                      tmp_map;
-          LevelInd                                        found_cell;
-
-          // Find the cells that the particles moved to
-          tmp_map.clear();
-          for (it=particles.begin(); it!=particles.end();)
-            {
-              found_cell = find_cell(it->second, it->first);
-              if (found_cell != it->first)
-                {
-                  tmp_map.insert(std::make_pair(found_cell, it->second));
-                  particles.erase(it++);
-                }
-              else ++it;
-            }
-          particles.insert(tmp_map.begin(),tmp_map.end());
-        };
+        template <class Archive>
+        void serialize (Archive &ar, const unsigned int version);
 
+      private:
```

Comment:
I tend to think that a whole lot of the functions above could actually be made `private` as well.